presto-main/src/main/java/com/facebook/presto/Session.java

```diff
@@ -30,16 +31,19 @@
 import java.util.Locale;
 import java.util.Map;
 import java.util.Map.Entry;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.TimeZone;
 
 import static com.google.common.base.MoreObjects.toStringHelper;
 import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
 import static java.util.Objects.requireNonNull;
 
 public final class Session
 {
     private final QueryId queryId;
+    private final Optional<TransactionId> transactionId;
```

Comment:
The main problem here is that we actually have 2 types of sessions now:
1) session submitted by client
2) session used for execution

The client can submit a session with a transaction or not. The execution also generally requires a transaction, but not in every case (e.g. when doing commit and rollback statements). I have a similar feeling about this as you, but I couldn't find a cleaner way to do it.